### PR TITLE
Send full host-id in CSR request

### PR DESCRIPTION
**Pull request by orenbm**
Connected to cyberark/conjur#1239

In order to authenticate hosts from every policy branch, we need the
authn-client to send the full host-id to the server, and in turn the server builds the conjur host-id
from the host-id given by the client instead of using the prefix {@account}:host:conjur/authn-k8s/#{@service_name}/apps.

We send the suffix in the common-name as we did before and we send the prefix
in the request's header.

### CODE REVIEW

File: pkg/authenticator/authenticator.go [hunk 1]
```diff
@@ -119,12 +113,21 @@ func (auth *Authenticator) Login() error {
 
 	log.InfoLogger.Printf(log.CAKC007I, auth.Config.Username)
 
-	csrRawBytes, err := auth.GenerateCSR()
+	// We send the hostname suffix in the CSR's common-name so we maintain backwards compatibility
```
[the remark this answers is not in the document]
Comment (orenbm) [reply pkg/authenticator/authenticator.go]: so it seems that the username of a host is `host/<host-id>`. I'll change all occurrences to `username` as here we have the `host/` prefix.

**Comment by orenbm** [comment]
@jvanderhoof @jonahx please re-review.

I added UTs and comments for visibility.

I didn't write this as a comment (didn't see it fit) but i wrote about it in the commit:
- Integration tests for this change are located in:
  - Conjur server: We deploy Conjur in GKE and send `/inject_client_cert`
                   and `/authenticate` requests in different permutations.
  - kubernetes-conjur-demo: We deploy a Conjur cluster and an `authn-client`
                            pod to see hosts can authenticate with it.

**Comment by orenbm** [comment]
@jvanderhoof can this PR get merged? I addressed all of @jonahx's requests.

File: pkg/authenticator/authenticator.go [hunk 2]
```diff
@@ -119,12 +112,23 @@ func (auth *Authenticator) Login() error {
 
 	log.InfoLogger.Printf(log.CAKC007I, auth.Config.Username)
 
-	csrRawBytes, err := auth.GenerateCSR()
+	// We want to let hosts authenticate with Conjur from anywhere in the policy
```
Comment (Tovli) [inline pkg/authenticator/authenticator.go]: Seems this comment should be in the ticket (which number should be part of this PR name)

File: pkg/authenticator/authenticator.go [hunk 3]
```diff
@@ -119,12 +112,23 @@ func (auth *Authenticator) Login() error {
...
+	// tree. For this to work, we send the full username in the CSR request, instead
+	// of just the username suffix (machine identity) like it was before.
+	// To maintain backwards compatibility, the CSR's common-name still contains
+	// only the username suffix (the machine identity) and the prefix is sent in
+	// a header named "Host-Id-Prefix". This header content will be extracted in
+	// the server and will be appended to the suffix from the CSR's common-name.
+	username := NewUsername(auth.Config.Username)
+
+	// Generate CSR
```
Comment (Tovli) [inline pkg/authenticator/authenticator.go]: why is this comment necessary?

File: pkg/authenticator/authenticator_test.go [hunk 4]
```diff
@@ -48,4 +52,32 @@ func TestAuthenticator(t *testing.T) {
 			So(authn.IsCertExpired(), ShouldEqual, true)
 		})
 	})
+
+	Convey("GenerateCSR", t, func() {
```
Comment (Tovli) [inline pkg/authenticator/authenticator_test.go]: Are those 3 different tests in one?
Why not creating 3 different tests?

File: pkg/authenticator/username.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
+package authenticator
+
+import "strings"
+
+// Represents the username of the host that is authenticating with Conjur.
```
Comment (Tovli) [inline pkg/authenticator/username.go]: Seems better to put this in the ticket

File: pkg/authenticator/username_test.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package authenticator
+
+import (
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestUsername(t *testing.T) {
```
Comment (Tovli) [inline pkg/authenticator/username_test.go]: Is there a standard for tests names? Like UsernameSplittedSuccessfully ?
What can I know if I see TestUsername failed in the output of the tests?

File: pkg/authenticator/authenticator.go (hunk 2 above)
Comment (orenbm) [reply pkg/authenticator/authenticator.go]: This comment may be also in the ticket but we also have it in the code for the next reader who will never look for a ticket.

Regarding the PR name - we do not have the ticket in it but we do have it in the description. I'll add it.

File: pkg/authenticator/authenticator.go (hunk 3 above)
Comment (orenbm) [reply pkg/authenticator/authenticator.go]: It is not. I will remove it.

File: pkg/authenticator/authenticator_test.go (hunk 4 above)
Comment (orenbm) [reply pkg/authenticator/authenticator_test.go]: it is one test only. 
`Convey("GenerateCSR"` states the beginning of the UT for `GenerateCSR`. Then we have the preparation, and then we have 1 `Given` clause and 2 `Then` clauses which test 2 outcomes:
1. The method finishes without an error
2. The common-name is inserted in the certificate subject as expected.

File: pkg/authenticator/username.go (hunk 5 above)
Comment (orenbm) [reply pkg/authenticator/username.go]: Again, this improves code readability.

File: pkg/authenticator/username_test.go (hunk 6 above)
Comment (orenbm) [reply pkg/authenticator/username_test.go]: Here is when the Convey clauses below come to action. For example, you will see a failure in:
`NewUsername -> Given a host's username -> Splits the username as expected` and will understand the context and the unexpected output.

File: pkg/authenticator/username_test.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	Convey("NewUsername", t, func() {
```
Comment (nessiLahav) [inline pkg/authenticator/username_test.go]: I think we should elaborate the testing here, and to understand which behavior we are expecting, for example:
1. username = string1
2. username = string1/string2

File: pkg/authenticator/requests_test.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+)
+
+func TestRequests(t *testing.T) {
+	Convey("LoginRequest", t, func() {
+		conjurVersion := "5"
+		authnURL := "dummyURL"
+		csrBytes := []byte("dummyCSRBytes")
+
+		Convey("Given a host's username prefix", func() {
+			usernamePrefix := "host.path.to.policy"
```
Comment (nessiLahav) [inline pkg/authenticator/requests_test.go]: Do you see any value to test also an empty Prefix ?

File: pkg/authenticator/username_test.go (hunk 7 above)
Comment (orenbm) [reply pkg/authenticator/username_test.go]: thanks, added more tests

File: pkg/authenticator/requests_test.go (hunk 8 above)
Comment (orenbm) [reply pkg/authenticator/requests_test.go]: we added input validation on the username that it includes a `host/` prefix so we don't have an empty prefix scenario

**Pull request merged**